src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java

```diff
@@ -539,12 +549,51 @@ public long importKey(long hSession, CK_ATTRIBUTE[] attributes) throws PKCS11Exc
             if (attr.type == CKA_KEY_TYPE) {
                 keyType = attr.getLong();
             }
-            if (attr.type == CKA_VALUE) {
-                keyBytes = attr.getByteArray();
+            ckAttrsMap.put(attr.type, attr);
+        }
+
+        if(keyClass == CKO_PRIVATE_KEY) {
+            if(keyType == CKK_RSA) {
+                try {
+                    keyBytes = RSAPrivateCrtKeyImpl.newKey(
+                        KeyType.RSA,
+                        null,
+                        ckAttrsMap.getOrDefault(CKA_MODULUS, new CK_ATTRIBUTE(CKA_MODULUS, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PUBLIC_EXPONENT, new CK_ATTRIBUTE(CKA_PUBLIC_EXPONENT, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PRIVATE_EXPONENT, new CK_ATTRIBUTE(CKA_PRIVATE_EXPONENT, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PRIME_1, new CK_ATTRIBUTE(CKA_PRIME_1, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PRIME_2, new CK_ATTRIBUTE(CKA_PRIME_2, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_EXPONENT_1, new CK_ATTRIBUTE(CKA_EXPONENT_1, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_EXPONENT_2, new CK_ATTRIBUTE(CKA_EXPONENT_2, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_COEFFICIENT, new CK_ATTRIBUTE(CKA_COEFFICIENT, BigInteger.ZERO)).getBigInteger()
+                    ).getEncoded();
+                } catch (InvalidKeyException e) {
+                    throw new PKCS11Exception(0x00000005L, null);
+                }
+            } else if (keyType == CKK_EC) {
+                try {
+                    keyBytes = ECUtil.generateECPrivateKey(
+                        ckAttrsMap.getOrDefault(CKA_VALUE, new CK_ATTRIBUTE(CKA_VALUE, BigInteger.ZERO)).getBigInteger(),
+                        ECUtil.getECParameterSpec(
+                            Security.getProvider("SunEC"),
+                            ckAttrsMap.get(CKA_EC_PARAMS).getByteArray())
```

Comment:
What guarantees that `CKA_EC_PARAMS` exists?

Reply:
As far as I know, an Elliptic Curve Private Key Object Attributes contains both `CKA_EC_PARAMS` and `CKA_VALUE`. The CKA_EC_PARAMS attribute of a private key which must be provided when the private key object is created.

Reply:
Yes, given this is predicated on `keyClass == CKO_PRIVATE_KEY`, I should have referred to `generatePrivate()` which does provide `CKA_EC_PARAMS`.

Another path that leads here via `C_CreateObject()` is from `P11KeyStore.storePkey()`, but that seems to provide the necessary attributes as well.

This seems rather "delicate" to me: perhaps an explicit check is warranted here that `CKA_EC_PARAMS` is available (throwing a `PKCS11Exception` if not).